org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ProcessTaskManager.java

```diff
@@ -46,159 +51,131 @@ public class ProcessTaskManager {
 		return t;
 	});
 
-public ProcessTaskManager(Compiler compiler, int startingIndex) {
-	this.compiler = compiler;
-	this.startingIndex = startingIndex;
+	public ProcessTaskManager(Compiler compiler, int startingIndex) {
+		this.compiler = compiler;
+		this.startingIndex = startingIndex;
 
-	this.currentIndex = 0;
-	this.availableIndex = 0;
-	this.size = PROCESSED_QUEUE_SIZE;
-	this.sleepCount = 0; // 0 is no one, +1 is the processing thread & -1 is the writing/main thread
-	this.units = new CompilationUnitDeclaration[this.size];
+		this.units = new ArrayBlockingQueue<>(PROCESSED_QUEUE_SIZE);
 
-	synchronized (this) {
-		this.processingTask = executor.submit(this::compile);
+		this.processingTask = executor.submit(this::processing);
 	}
-}
 
 // add unit to the queue - wait if no space is available
-private synchronized void addNextUnit(CompilationUnitDeclaration newElement) {
-	while (this.units[this.availableIndex] != null) {
-		//System.out.print('a');
-		//if (this.sleepCount < 0) throw new IllegalStateException(Integer.valueOf(this.sleepCount).toString());
-		this.sleepCount = 1;
+	private void addNextUnit(Object newElement) {
 		try {
-			wait(250);
-		} catch (InterruptedException ignore) {
-			// ignore
+			this.units.put(newElement);
+		} catch (InterruptedException interrupt) {
+			throw new RuntimeException(interrupt);
```

Comment:
1) Why is this wrapped? Usually if interrupted exception is handled, it should be flagged by Thread.interrupted() call, but neither this code nor caller does that.

2) Shouldn't this be AbortCompilation exception if we encounter problems during queue writing?

Reply:
InterruptedException is a checked exception. that is not supported by executor.submit().
By now nothing should try to interrupt the executor, so it would be an error to receive one!
All other places in JDT just ignore InterruptedException. we could do that here too but that would need an endless loop until not interrupted. AbortCompilation is not supported here either as nothing reads the future. Only close() does - but that happens anyway only at the end.